pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -67,49 +67,59 @@ This file implements string parsing and creation for NumPy datetime.
  * Returns 0 on success, -1 on failure.
  */
 
+enum Comparison {
+    ADVANCE,
+    RETURN,
+    ERROR
+};
 // This function will advance the pointer on format
 // and decrement characters_remaining by n on success
-// On failure will return -1 without incrementing
-static int compare_format(const char **format, int *characters_remaining,
-                 const char *compare_to, int n, const int exact) {
+// On failure will return ERROR without incrementing
+// If `exact` is PARTIAL_MATCH, and the `format` string has
+// been exhausted, then signal to the caller to finish parsing.
+static enum Comparison compare_format(
+        const char **format,
+        int *characters_remaining,
+        const char *compare_to,
+        int n,
+        const enum Exact exact
+) {
+  if (exact == PARTIAL_MATCH && !*characters_remaining) {
```

Comment:
```suggestion
  if (exact == PARTIAL_MATCH && *characters_remaining == 0) {
```

Nit but would be good to explicitly compare to 0. Depending on code structure we may also want to be careful what happens if `characters_remaining` somehow ends up as negative